node/core/backing/src/lib.rs

```diff
@@ -0,0 +1,1149 @@
+
+		(
+			Self {
+				keystore,
+				parent,
+				rx_to,
+				tx_to,
+				tx_from,
+
+				validation_code: ValidationCode::default(),
+				head_data: HeadData::default(),
+				assignment: ParaId::default(),
+
+				issued_validity: HashSet::default(),
+				backed_candidates: HashSet::default(),
+				seconded: None,
+				backing_watchers: Vec::new(),
+				reported_misbehavior_for: HashSet::default(),
+
+				table: Table::default(),
+				table_context: TableContext::default(),
+			},
+			rx_from,
+		)
+	}
+
+	/// Hand out the sending part of the channel to communicate with this job.
+	fn tx(&mut self) -> mpsc::Sender<ToJob> {
+		self.tx_to.clone()
+	}
+
+	/// Request a validator set from the `RuntimeApi`.
+	async fn request_validators(&mut self) -> Result<Vec<ValidatorId>, Error> {
+		let (tx, rx) = oneshot::channel();
+
+		self.tx_from.send(FromJob::RuntimeApiMessage(RuntimeApiMessage::Request(
+				self.parent.clone(),
+				RuntimeApiRequest::Validators(tx),
+			)
+		)).await?;
+
+		Ok(rx.await?)
+	}
+
+	/// Request the scheduler roster from `RuntimeApi`.
+	async fn request_validator_groups(&mut self) -> Result<SchedulerRoster, Error> {
+		let (tx, rx) = oneshot::channel();
+
+		self.tx_from.send(FromJob::RuntimeApiMessage(RuntimeApiMessage::Request(
+				self.parent.clone(),
+				RuntimeApiRequest::ValidatorGroups(tx),
+			)
+		)).await?;
+
+		Ok(rx.await?)
+	}
+
+	/// Request the validation code for some `ParaId` at given block
+	/// from `RuntimeApi`.
+	async fn request_validation_code(
+		&mut self,
+		id: ParaId,
+		parent: BlockNumber,
+		parablock: Option<BlockNumber>,
+	) -> Result<ValidationCode, Error> {
+		let (tx, rx) = oneshot::channel();
+
+		self.tx_from.send(FromJob::RuntimeApiMessage(RuntimeApiMessage::Request(
+				self.parent.clone(),
+				RuntimeApiRequest::ValidationCode(id, parent, parablock, tx),
+			)
+		)).await?;
+
+		Ok(rx.await?)
+	}
+
+	/// Request a `SigningContext` from the `RuntimeApi`.
+	async fn request_signing_context(&mut self) -> Result<SigningContext, Error> {
+		let (tx, rx) = oneshot::channel();
+
+		self.tx_from.send(FromJob::RuntimeApiMessage(RuntimeApiMessage::Request(
+				self.parent.clone(),
+				RuntimeApiRequest::SigningContext(tx),
+			)
+		)).await?;
+
+		Ok(rx.await?)
+	}
+
+	/// Request `HeadData` for some `ParaId` from `RuntimeApi`.
+	async fn request_head_data(&mut self, id: ParaId) -> Result<HeadData, Error> {
+		let (tx, rx) = oneshot::channel();
+
+		self.tx_from.send(FromJob::RuntimeApiMessage(RuntimeApiMessage::Request(
+				self.parent.clone(),
+				RuntimeApiRequest::HeadData(id, tx),
+			)
+		)).await?;
+
+		Ok(rx.await?)
+	}
+
+	/// Logic that runs on startup.
+	async fn on_startup(&mut self) -> Result<(), Error> {
+		let validators = self.request_validators().await?;
+	
+		let roster = self.request_validator_groups().await?; 
+
+		self.table_context.key = signing_key(&validators[..], &self.keystore);
+
+		for assignment in roster.scheduled {
```

Comment:
Also, at some point (not in this PR) we will want to extend the functionality of the runtime API so that we can get a new `roster` contingent on availability bitfields that we've seen. I'm not sure exactly how that will fit in. But this will be necessary in order to avoid having a max-throughput of 1 parablock per 2 relay-chain blocks.